File: go/tromp.c
/* John Tromp's program (comments Jake Hennig and rbh 2024) */
/* C program to solve 2x2 go with area rules and positional superko */
/* note that suicide is not possible in this case */
/* trying moves before passes slows search dramatically */

#include <stdio.h>
#define NSHOW 4  /* output of search displayed to this depth */
#define NMOVES 4 /* max number of non-pass moves */
#define CUT 1    /* CUT 1: pruning, CUT 0: minimax */

char h[256];    /* bitmap of positions in game history */
int nodes[99];  /* number nodes visited at each depth */
long ngames;    /* total number games played */

void show(int n, int black, int white, int alpha, int beta, int passed) /* position's 2-line ASCII rep'n */
{
  int i;
  printf("%d (%d,%d)%s\n", n, alpha, beta, passed ? " pass" : ""); /* depth, alpha, beta, whether player passed */
  for (i = 0; i < 4; i++) {
    printf(" %c", ".XO#"[(black >> i & 1) + 2 * (white >> i & 1)]); /* boardstate (X-black O-white) */
    if (i & 1)
      putchar('\n'); /* 2x2 board, so new_line after each two characters */
  }
} /* 1000 top-left  0100 top-right  0010 bottom-left  0001 bottom-right*/

void visit(int black, int white)
{
  h[black + 16 * white] = 1; /* mark position visited in game history bitmap */
}

void unvisit(int black, int white)
{
  h[black + 16 * white] = 0; /* mark position unvisited in game histroy bitmap */
}

int visited(int black, int white)
{
  return h[black + 16 * white]; /* whether position already visited */
}

int owns(int bb)
{
  return bb == (1 | 8) || bb == (2 | 4); /* whether player owns both board corners */
}

int popcnt[16] = {0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4}; /* number 1-bits in bin-rep, numbers 0..15 */
/* 0000, 0001, 0010, 0011, 0100, 0101... 1110, 1111 */
/* used to count stones of each player */

int score(int black, int white)
{
  ngames++; /* number of games played */
  if (black == 0)          /* no black stones */
    return white ? -4 : 0; /* white wins (-4) unless also no white stones (draw: 0) */
  if (white == 0)          /* no white stones */
    return 4;              /* black wins (+4) */
  else
    return popcnt[black] - popcnt[white]; /* score is diff in stone count */
}

int xhasmove(int black, int white, int move, int *newblack, int *newwhite) /* whether black has valid move */
{
  move = 1 << move; /* shift move left 1 bit */
  /*  position already occupied? black has 3 stones? white owns board?  */
  if ((black | white) & move || popcnt[black] == 3 || owns(white)) 
    return 0; /* move invalid */
  *newblack = black | move; /* updates black position after move */
  *newwhite = (*newblack | white) == 15 || owns(*newblack) ? 0 : white; /* updates white position after move */
  return !visited(*newblack, *newwhite); /* whether position already visited */
}

int ohasmove(int black, int white, int move, int *newblack, int *newwhite) /* see above: xhasmove() */
{
  move = 1 << move; 
  if ((black | white) & move || popcnt[white] == 3 || owns(black)) 
    return 0; 
  *newwhite = white | move; 
  *newblack = (*newwhite | black) == 15 || owns(*newwhite) ? 0 : black; 
  return !visited(*newblack, *newwhite); 
}

int xab(int n, int black, int white, int alpha, int beta, int passed) /* black alphabeta search */
{
  int i, s, newblack, newwhite;
  int oab(int, int, int, int, int, int); /* forward declaration for white ab-search */

  nodes[n]++; /* nodes visited at this depth */
  if (n < NSHOW)
    show(n, black, white, alpha, beta, passed); /* display position if within NSHOW depth */

  s = passed ? score(black, white) : oab(n + 1, black, white, alpha, beta, 1); 
  /* if opponent passed and we pass then calculate score, otw call oab() */
  /* whether score greater than alpha (update alpha if necessary)
     and whether alpha >= beta and whether CUT flag set */
  if (s > alpha && (alpha = s) >= beta && CUT)
    /* prune if current score > alpha and after update whether alpha still >= beta */
    return alpha;

  for (i = 0; i < NMOVES; i++) { /* loop over moves */
    if (xhasmove(black, white, i, &newblack, &newwhite)) { /* whether move valid */
      visit(newblack, newwhite); 
      s = oab(n + 1, newblack, newwhite, alpha, beta, 0); /* explore opponent response */
      unvisit(newblack, newwhite); 
      if (s > alpha && (alpha = s) >= beta && CUT)
        return alpha; /* same as above */
    }
  }

  /* search pass move last? move lines 91-98 here */

  return alpha;
}

int oab(int n, int black, int white, int alpha, int beta, int passed) /* see above xab() */
{
  int i, s, newblack, newwhite;

  nodes[n]++; 
  if (n < NSHOW)
    show(n, black, white, alpha, beta, passed); 

  s = passed ? score(black, white) : xab(n + 1, black, white, alpha, beta, 1); 
  if (s < beta && (beta = s) <= alpha && CUT)
    return beta; 

  for (i = 0; i < NMOVES; i++) { 
    if (ohasmove(black, white, i, &newblack, &newwhite)) { 
      visit(newblack, newwhite); 
      s = xab(n + 1, newblack, newwhite, alpha, beta, 0); 
      unvisit(newblack, newwhite); 
      if (s < beta && (beta = s) <= alpha && CUT)
        return beta; 
    }
  }

  /* search pass move last? move lines 122-125 here */

  return beta;
}

int main()
{
  int i, c, s;
  c = xab(0, 0, 0, -4, 4, 0); /* start alphabeta search from empty board */
  for (i = s = 0; nodes[i]; i++) {
    s += nodes[i]; /* total nodes visited */
    printf("%d: %d\n", i, nodes[i]); /* nodes visited at each depth */
  }
  printf("total: %d\nngames: %ld\nx wins by %d\n", s, ngames, c); /* total nodes visited and game stats */
  return 0;
}

File: go/tromp.py
# Jake Hennig 2024
# translation of John Tromp's c program to python3
# solve 2x2 go with area rules and positional superko

# Note that suicide is not possible in this case
# Trying moves before passes slows search dramatically

NSHOW = 4  # max depth for displying search output
NMOVES = 4  # max number possible moves
CUT = 1  # CUT 1 allows pruning otherwise, CUT 0 is minimax

h = [0] * 256  # bitmap of positions in game history
nodes = [0] * 99  # number of nodes visited at each depth
ngames = 0  # total  number games played

def show(n, black, white, alpha, beta, passed):
    """Print 2-line ASCII representation of position"""
    print(f"{n} ({alpha},{beta})", "pass" if passed else "") # prints depth, alpha, beta, and if player passed turn */
    for i in range(4):
        if (i % 2 == 0):
            print(" ", end="")
        print(".XO#"[((black >> i) & 1) + 2 * ((white >> i) & 1)], end=" ") # prints the board state (X - black, O - white) */
        if (i & 1):
            print() # place new line after every two characters - 2x2 board */
# 1000 top-left, 0100 top-right, 0010 bottom-left, 0001 bottom-right

def visit(black, white):
    """Mark the position as visited"""
    h[black + 16 * white] = 1

def unvisit(black, white):
    """Mark the position as unvisited"""
    h[black + 16 * white] = 0

def visited(black, white):
    """Check if the position has been visited"""
    return h[black + 16 * white]

def owns(bb):
    """Check if one player has both corners of the board (control over it)"""
    return bb == (1 | 8) or bb == (2 | 4)

popcnt = [0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4]  # number of 1 bits in the binary representation of numbers from 0 to 15
# 0000, 0001, 0010, 0011, 0100, 0101... 1110, 1111
# Used to count stones of each player

def score(black, white):
    """Calculate the score of the position"""
    global ngames
    ngames += 1 # counts number of games played
    if (black == 0): return -4 if white else 0 # black no stones? white wins (-4) unless white also no stones (draw 0)
    if (white == 0): return 4 # white no stones: black wins +4 */
    else: return popcnt[black] - popcnt[white] # score is stones difference

def xhasmove(black, white, move):
    """Check if black has a valid move"""
    move = 1 << move # get move binary representation
    if (black | white) & move or popcnt[black] == 3 or owns(white): return False # no 
    newblack = black | move # update black position
    newwhite = 0 if (newblack | white) == 15 or owns(newblack) else white # update white position after move
    return not visited(newblack, newwhite) # new position visited before?

def ohasmove(black, white, move): # see xhasmove for comments
    """Check if white has a valid move"""
    move = 1 << move 
    if (black | white) & move or popcnt[white] == 3 or owns(black): return False 
    newwhite = white | move 
    newblack = 0 if (newwhite | black) == 15 or owns(newwhite) else black 
    return not visited(newblack, newwhite) 

def xab(n, black, white, alpha, beta, passed):
    """Alpha-beta search for black's turn"""
    global nodes
    nodes[n] += 1 # nodes visited at this depth
    if n < NSHOW: show(n, black, white, alpha, beta, passed) # displays board state if within NSHOW depth

    # if opponent passed and we now pass, terminal position so calculate score, otw continue search
    s = score(black, white) if passed else oab(n + 1, black, white, alpha, beta, 1) 
    if (s > alpha):
        alpha = s
        if (alpha >= beta and CUT): return alpha # prune if score  > alpha and after update whether alpha  >= beta

    for i in range(NMOVES): # loop through possible moves
        if (xhasmove(black, white, i)):
            newblack, newwhite = black, white
            move = 1 << i
            newblack = black | move
            newwhite = 0 if (newblack | white) == 15 or owns(newblack) else white
            visit(newblack, newwhite) 
            s = oab(n + 1, newblack, newwhite, alpha, beta, 0)
            unvisit(newblack, newwhite) 
            if (s > alpha):
                alpha = s
                if (alpha >= beta and CUT): return alpha # prune if score > alpha and after update whether alpha >= beta
    return alpha

def oab(n, black, white, alpha, beta, passed): # see xab for comments
    """Alpha-beta search for white's turn"""
    global nodes
    nodes[n] += 1 
    if (n < NSHOW): show(n, black, white, alpha, beta, passed) 

    s = score(black, white) if passed else xab(n + 1, black, white, alpha, beta, 1) 
    if (s < beta):
        beta = s
        if (beta <= alpha and CUT): return beta 

    for i in range(NMOVES): 
        if (ohasmove(black, white, i)): 
            newblack, newwhite = black, white
            move = 1 << i
            newwhite = white | move
            newblack = 0 if (newwhite | black) == 15 or owns(newwhite) else black
            visit(newblack, newwhite) 
            s = xab(n + 1, newblack, newwhite, alpha, beta, 0)
            unvisit(newblack, newwhite) 
            if (s < beta):
                beta = s
                if (beta <= alpha and CUT): return beta 
    return beta

def main():
    s = 0
    c = xab(0, 0, 0, -4, 4, 0) # start alphabeta search from empty board
    for i, count in enumerate(nodes):
        s += count # total nodes visited
        if (count):
            print(f"{i}: {count}") # nodes visited at each depth

    print(f"total: {s}\nngames: {ngames}\nx wins by {c}") # nodes visited and game stats

if __name__ == "__main__":
    main()
